icu4j/main/core/src/main/java/com/ibm/icu/text/UnicodeSet.java

```diff
@@ -5127,5 +5143,179 @@ public static void setDefaultXSymbolTable(XSymbolTable xSymbolTable) {
+     * Do not alter the {@code UnicodeSet} while iterating.
+     *
+     * @return a {@link Stream} of {@code String}
+     */
+    public Stream<String> stringStream() {
+        return StreamSupport.stream(strings().spliterator(), false);
+    }
+
+    /**
+     * Returns an {@link IntStream} of Unicode code point values from this {@code UnicodeSet}.
+     *
+     * <p><b>Warning:</b> To iterate over the full contents, you have to also iterate over the strings.
+     *
+     * <p><b>Warning:</b> For speed, {@code UnicodeSet} iteration does not check for concurrent modification.<br>
+     * Do not alter the {@code UnicodeSet} while iterating.
+     *
+     * @return an {@link IntStream} of Unicode code point values
+     */
+    public IntStream codePointStream() {
+        return StreamSupport.intStream(new CodePointsSpliterator(this), false);
+    }
+
+    /**
+     * Returns a stream of {@code String} values from this {@code UnicodeSet}.
+     *
+     * <p><b>Warning:</b> To iterate over the full contents, you have to also iterate over the strings.
+     *
+     * <p><b>Warning:</b> For speed, {@code UnicodeSet} iteration does not check for concurrent modification.<br>
+     * Do not alter the {@code UnicodeSet} while iterating.
+     *
+     * @return a {@link Stream} of {@code String}
+     */
+    public Stream<String> stream() {
+        return StreamSupport.stream(spliterator(), false);
+    }
+
+    /**
+     * Provide for iteration on code points.
+     *
+     * <p>Returns an {@link Iterable} over all the code points, "expanding" the ranges
+     * by iterating on all code points between the {@link EntryRange#codepoint} and
+     * {@link EntryRange#codepointEnd}.
+     *
+     * <p><b>Warning:</b> This is a convenience method, but comes with a performance penalty
+     * because it boxes {@code int} into {@code Integer}.<br>
+     * For faster iteration use the {@link #codePointStream()} or {@link #charAt(int)}.
+     *
+     * <p><b>Warning:</b> To iterate over the full contents, you have to also iterate over the strings.
+     *
+     * <p><b>Warning:</b> For speed, {@code UnicodeSet} iteration does not check for concurrent modification.<br>
+     * Do not alter the {@code UnicodeSet} while iterating.
+     *
+     * @return an {@link Iterable} over all the code points
+     */
+    public Iterable<Integer> codePoints() {
+        return new CodePointsIterable();
+    }
+
+    private class CodePointsIterable implements Iterable<Integer> {
+        @Override
+        public Iterator<Integer> iterator() {
+            return new CodePointsIterator();
+        }
+    }
+
+    private class CodePointsIterator implements Iterator<Integer> {
+        private int pos = 0;
+        private int lastCp = -1; // don't call `charAt` twice, in `hasNext()` and `next()`
+
+        @Override
+        public boolean hasNext() {
+            lastCp = charAt(pos++);
+            return lastCp != -1;
+        }
```

Comment:
1. I wouldn't iterate in hasNext(). I would init the iterator with the number of code points or ranges, keep and index, and make hasNext() compare the index with the total number.
2. Don't call charAt(index). It's slow.
3. I suggest you init with getRangeCount(), keep a rangeIndex, and when you go to the next range then call getRangeStart(rangeIndex) and getRangeEnd(rangeIndex); store those, iterate over each code point of the current range.

Reply:
Done

Refactored all the code to work directly on the list and not calling `charAt`